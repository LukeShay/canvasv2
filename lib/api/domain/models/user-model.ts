/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable global-require */
import { Model, JSONSchema, RelationMappings } from 'objection';
import { IUser, UserRole } from '../../../types';
import Constants from '../constants';

export class UserModel extends Model implements IUser {
  address1?: string | undefined;

  address2?: string | undefined;

  city?: string | undefined;

  createdAt!: string;

  email!: string;

  firstName!: string;

  id!: string;

  lastName!: string;

  password!: string;

  role!: UserRole;

  state?: string | undefined;

  updatedAt!: string;

  zip?: string | undefined;

  static tableName = Constants.USERS_TABLE;

  static idColumn = 'id';

  static get jsonSchema(): JSONSchema {
    return {
      properties: {
        addresss1: { type: ['string', 'null'] },
        addresss2: { type: ['string', 'null'] },
        city: { type: ['string', 'null'] },
        createdAt: { type: 'timestamp' },
        email: { type: 'string' },
        firstName: { maxLength: 255, minLength: 1, type: 'string' },
        id: { type: 'uuid' },
        lastName: { maxLength: 255, minLength: 1, type: 'string' },
        state: { type: ['string', 'null'] },
        stateId: { type: ['string', 'null'] },
        updatedAt: { type: 'timestamp' },
        zip: { type: ['string', 'null'] },
      },
      required: ['email', 'firstName', 'lastName', 'password'],
      type: 'object',
    };
  }

  static get relationMappings(): RelationMappings {
    const { ClassModel } = require('./class-model');
    const { AssistantModel } = require('./assistant-model');
    const { OrganizationModel } = require('./organization-model');
    const { StateModel } = require('./state-model');
    const { OrganizationMemberModel } = require('./organization-member-model');
    const { StudentModel } = require('./student-model');

    return {
      assistantClasses: {
        join: {
          from: `${UserModel.tableName}.${UserModel.idColumn}`,
          through: {
            from: `${AssistantModel.tableName}.studentId`,
            to: `${AssistantModel.tableName}.classId`,
          },
          to: `${ClassModel.tableName}.${ClassModel.idColumn}`,
        },
        modelClass: ClassModel,
        relation: Model.ManyToManyRelation,
      },
      organizations: {
        join: {
          from: `${UserModel.tableName}.${UserModel.idColumn}`,
          through: {
            from: `${OrganizationMemberModel.tableName}.userId`,
            to: `${OrganizationMemberModel.tableName}.organizationId`,
          },
          to: `${OrganizationModel.tableName}.${OrganizationModel.idColumn}`,
        },
        modelClass: OrganizationModel,
        relation: Model.ManyToManyRelation,
      },
      state: {
        join: {
          from: `${UserModel.tableName}.stateId`,
          to: `${StateModel.tableName}.${StateModel.idColumn}`,
        },
        modelClass: StateModel,
        relation: Model.BelongsToOneRelation,
      },
      studentClasses: {
        join: {
          from: `${UserModel.tableName}.${UserModel.idColumn}`,
          through: {
            from: `${StudentModel.tableName}.studentId`,
            to: `${StudentModel.tableName}.classId`,
          },
          to: `${ClassModel.tableName}.${ClassModel.idColumn}`,
        },
        modelClass: ClassModel,
        relation: Model.ManyToManyRelation,
      },
    };
  }
}
