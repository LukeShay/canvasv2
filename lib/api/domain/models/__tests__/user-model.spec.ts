import { UserModel } from '../user-model';

describe('UserModel', () => {
  it("should have idColum 'id'", () => {
    expect(UserModel.idColumn).toEqual('id');
  });

  it('should should have jsonSchema that matches snapshot', () => {
    expect(UserModel.jsonSchema).toMatchSnapshot();
  });

  it('should should have relationMappings that matches snapshot', () => {
    expect(UserModel.relationMappings).toMatchSnapshot();
  });
});
