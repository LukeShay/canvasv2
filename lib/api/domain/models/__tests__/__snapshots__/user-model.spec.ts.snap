// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserModel should should have jsonSchema that matches snapshot 1`] = `
Object {
  "properties": Object {
    "addresss1": Object {
      "type": Array [
        "string",
        "null",
      ],
    },
    "addresss2": Object {
      "type": Array [
        "string",
        "null",
      ],
    },
    "city": Object {
      "type": Array [
        "string",
        "null",
      ],
    },
    "createdAt": Object {
      "type": "timestamp",
    },
    "email": Object {
      "type": "string",
    },
    "firstName": Object {
      "maxLength": 255,
      "minLength": 1,
      "type": "string",
    },
    "id": Object {
      "type": "uuid",
    },
    "lastName": Object {
      "maxLength": 255,
      "minLength": 1,
      "type": "string",
    },
    "state": Object {
      "type": Array [
        "string",
        "null",
      ],
    },
    "stateId": Object {
      "type": Array [
        "string",
        "null",
      ],
    },
    "updatedAt": Object {
      "type": "timestamp",
    },
    "zip": Object {
      "type": Array [
        "string",
        "null",
      ],
    },
  },
  "required": Array [
    "email",
    "firstName",
    "lastName",
    "password",
  ],
  "type": "object",
}
`;

exports[`UserModel should should have relationMappings that matches snapshot 1`] = `
Object {
  "assistantClasses": Object {
    "join": Object {
      "from": "Users.id",
      "through": Object {
        "from": "Assistants.studentId",
        "to": "Assistants.classId",
      },
      "to": "Classes.id",
    },
    "modelClass": [Function],
    "relation": [Function],
  },
  "organizations": Object {
    "join": Object {
      "from": "Users.id",
      "through": Object {
        "from": "Assistants.userId",
        "to": "Assistants.organizationId",
      },
      "to": "Assistants.id",
    },
    "modelClass": [Function],
    "relation": [Function],
  },
  "state": Object {
    "join": Object {
      "from": "Users.stateId",
      "to": "States.id",
    },
    "modelClass": [Function],
    "relation": [Function],
  },
  "studentClasses": Object {
    "join": Object {
      "from": "Users.id",
      "through": Object {
        "from": "Students.studentId",
        "to": "Students.classId",
      },
      "to": "Classes.id",
    },
    "modelClass": [Function],
    "relation": [Function],
  },
}
`;
